From 2932c4c1064689b50d7b9c9204625164a4d34b83 Mon Sep 17 00:00:00 2001
From: Jeffy Chen <jeffy.chen@rock-chips.com>
Date: Wed, 27 Feb 2019 16:52:15 +0800
Subject: [PATCH 4/9] qwaylandxdgshellv6: Support setting window position

Support setting window position.
Note:
1/ (0,0) initial position(default position) would be ignored.
2/ The decoration would be ignored when the space not enough.
3/ QT would not aware of the wayland position.

Change-Id: Ifb1433b3902d44c1b2e43036bc1805a6e00128fb
Signed-off-by: Jeffy Chen <jeffy.chen@rock-chips.com>
---
 .../xdg-shell-v6/qwaylandxdgshellv6.cpp           | 15 +++++++++++++++
 .../xdg-shell-v6/qwaylandxdgshellv6_p.h           |  2 ++
 2 files changed, 17 insertions(+)

diff --git a/src/plugins/shellintegration/xdg-shell-v6/qwaylandxdgshellv6.cpp b/src/plugins/shellintegration/xdg-shell-v6/qwaylandxdgshellv6.cpp
index dc3cfdb..6bc2a14 100644
--- a/src/plugins/shellintegration/xdg-shell-v6/qwaylandxdgshellv6.cpp
+++ b/src/plugins/shellintegration/xdg-shell-v6/qwaylandxdgshellv6.cpp
@@ -318,6 +318,21 @@ void QWaylandXdgSurfaceV6::propagateSizeHints()
 void QWaylandXdgSurfaceV6::setWindowGeometry(const QRect &rect)
 {
     set_window_geometry(rect.x(), rect.y(), rect.width(), rect.height());
+
+    if (m_window) {
+        QPoint position =
+            m_window->geometry().marginsRemoved(m_window->frameMargins()).topLeft();
+
+        // Also avoid initial position (0,0).
+        // What if we do want to be at (0,0)?
+        if (m_position == position)
+            return;
+        m_position = position;
+
+        // HACK: Set window position through .set_window_geometry(x, y, 0, 0)
+        set_window_geometry(position.x() > 0 ? position.x() : 0,
+                            position.y() > 0 ? position.y() : 0, 0, 0);
+    }
 }
 
 void QWaylandXdgSurfaceV6::setSizeHints()
diff --git a/src/plugins/shellintegration/xdg-shell-v6/qwaylandxdgshellv6_p.h b/src/plugins/shellintegration/xdg-shell-v6/qwaylandxdgshellv6_p.h
index f77a4d4..babe8a9 100644
--- a/src/plugins/shellintegration/xdg-shell-v6/qwaylandxdgshellv6_p.h
+++ b/src/plugins/shellintegration/xdg-shell-v6/qwaylandxdgshellv6_p.h
@@ -147,6 +147,8 @@ private:
     QRegion m_exposeRegion;
     uint m_pendingConfigureSerial = 0;
 
+    QPoint m_position;
+
     friend class QWaylandXdgShellV6;
 };
 
-- 
2.20.1

